include/deal.II/lac/petsc_ts.h

```diff
@@ -0,0 +1,706 @@
+        PetscFunctionReturn(0);
+      };
+
+      auto _dealii_ts_ijacobian_ = [](TS        ts,
+                                      PetscReal t,
+                                      Vec       x,
+                                      Vec       xdot,
+                                      PetscReal s,
+                                      Mat       A,
+                                      Mat       P,
+                                      void *    ctx) -> PetscErrorCode {
+        TimeStepper *myctx;
+
+        PetscFunctionBeginUser;
+        (void)ctx;
+        AssertTS(TSGetApplicationContext(ts, (void *)&myctx));
+        auto user = static_cast<TimeStepper *>(myctx);
+
+        VectorType  xdealii(x);
+        VectorType  xdotdealii(xdot);
+        AMatrixType Adealii(A);
+        PMatrixType Pdealii(P);
+        user->implicit_jacobian(t, xdealii, xdotdealii, s, Adealii, Pdealii);
+        PetscBool flg;
+        AssertTS(PetscObjectTypeCompare((PetscObject)A, MATMFFD, &flg));
+        if (flg)
+          {
+            AssertTS(MatAssemblyBegin(A, MAT_FINAL_ASSEMBLY));
+            AssertTS(MatAssemblyEnd(A, MAT_FINAL_ASSEMBLY));
+          }
+        PetscFunctionReturn(0);
+      };
+
+      auto _dealii_ts_ijacobian_with_setup_ = [](TS        ts,
+                                                 PetscReal t,
+                                                 Vec       x,
+                                                 Vec       xdot,
+                                                 PetscReal s,
+                                                 Mat       A,
+                                                 Mat       P,
+                                                 void *ctx) -> PetscErrorCode {
+        TimeStepper *myctx;
+
+        PetscFunctionBeginUser;
+        (void)ctx;
+        AssertTS(TSGetApplicationContext(ts, (void *)&myctx));
+        auto user = static_cast<TimeStepper *>(myctx);
+
+        VectorType  xdealii(x);
+        VectorType  xdotdealii(xdot);
+        AMatrixType Adealii(A);
+        PMatrixType Pdealii(P);
+
+        user->A = &Adealii;
+        user->P = &Pdealii;
+        user->setup_jacobian(t, xdealii, xdotdealii, s);
+
+        PetscBool flg;
+        AssertTS(PetscObjectTypeCompare((PetscObject)A, MATMFFD, &flg));
+        if (flg)
+          {
+            AssertTS(MatAssemblyBegin(A, MAT_FINAL_ASSEMBLY));
+            AssertTS(MatAssemblyEnd(A, MAT_FINAL_ASSEMBLY));
+          }
+        PetscFunctionReturn(0);
+      };
+
+      auto _dealii_ts_rhsfunction_ =
+        [](TS ts, PetscReal t, Vec x, Vec f, void *ctx) -> PetscErrorCode {
+        TimeStepper *myctx;
+
+        PetscFunctionBeginUser;
+        (void)ctx;
+        AssertTS(TSGetApplicationContext(ts, (void *)&myctx));
+        auto user = static_cast<TimeStepper *>(myctx);
+
+        VectorType xdealii(x);
+        VectorType fdealii(f);
+
+        user->explicit_function(t, xdealii, fdealii);
```

Comment:
Looking at this here, my sense is that `VectorType` really can only be a type that *wraps* a PETSc vector. It can't be a type that can be *initialized* by a PETSc `Vec` object, because if that were so, the results would not flow back into the `f` argument.

This ties back to my question of what kinds of template arguments are actually supported. I think we should document what the expectations are.

Reply:
done, take a look